File: backend_ruby/link_preview_service.rb
require "cgi"
require "ipaddr"
require "net/http"
require "resolv"
require "uri"
require_relative "api_error"

class LinkPreviewService
  MAX_PREVIEW_BYTES = 1_000_000

  def fetch(raw_url)
    current = validate_public_url(raw_url)
    4.times do
      request = Net::HTTP::Get.new(current)
      request["User-Agent"] = "Mozilla/5.0 (compatible; QiitaSearchLinkPreview/1.0)"
      request["Accept"] = "text/html,application/xhtml+xml"
      response = Net::HTTP.start(
        current.host, current.port,
        use_ssl: current.scheme == "https",
        open_timeout: 3,
        read_timeout: 5
      ) { |http| http.request(request) }

      if response.is_a?(Net::HTTPRedirection) && response["location"]
        current = validate_public_url(URI.join(current.to_s, response["location"]).to_s)
        next
      end
      raise ApiError.new(502, "リンク先の情報を取得できませんでした。") unless response.is_a?(Net::HTTPSuccess)

      content_type = response["content-type"].to_s.downcase
      unless content_type.include?("text/html") || content_type.include?("application/xhtml+xml")
        raise ApiError.new(400, "HTMLページではないためプレビューできません。")
      end
      document = response.body.byteslice(0, MAX_PREVIEW_BYTES).to_s
      title = first_non_empty(meta(document, "property", "og:title"), title_tag(document), current.host)
      description = first_non_empty(meta(document, "property", "og:description"), meta(document, "name", "description"))
      image = absolute_http_url(meta(document, "property", "og:image"), current)
      site_name = first_non_empty(meta(document, "property", "og:site_name"), current.host)
      return {
        url: current.to_s,
        title: title.each_char.take(300).join,
        description: description.each_char.take(500).join,
        image: image,
        site_name: site_name.each_char.take(100).join
      }
    end
    raise ApiError.new(502, "リンク先の情報を取得できませんでした。")
  rescue SocketError
    raise ApiError.new(400, "リンク先のホストを解決できません。")
  rescue Net::OpenTimeout, Net::ReadTimeout, Errno::ECONNREFUSED
    raise ApiError.new(502, "リンク先の情報を取得できませんでした。")
  end

  private

  def validate_public_url(raw)
    uri = URI(raw)
    unless %w[http https].include?(uri.scheme) && uri.host
      raise ApiError.new(400, "http または https のURLを指定してください。")
    end
    raise ApiError.new(400, "認証情報を含むURLはプレビューできません。") if uri.user || uri.password

    addresses = Resolv.getaddresses(uri.host)
    raise ApiError.new(400, "リンク先のホストを解決できません。") if addresses.empty?
    addresses.each do |address|
      ip = IPAddr.new(address)
      unless ip.ipv4? || ip.ipv6?
        raise ApiError.new(400, "リンク先のホストを解決できません。")
      end
      if ip.private? || ip.loopback? || ip.link_local? || ip.to_s == "0.0.0.0" || ip.to_s == "::"
        raise ApiError.new(400, "ローカルネットワークのURLはプレビューできません。")
      end
    end
    uri
  rescue URI::InvalidURIError
    raise ApiError.new(400, "http または https のURLを指定してください。")
  end

  def meta(document, attribute, value)
    quoted = Regexp.escape(value)
    patterns = [
      /<meta[^>]+#{attribute}=["']#{quoted}["'][^>]+content=["']([^"']*)["'][^>]*>/im,
      /<meta[^>]+content=["']([^"']*)["'][^>]+#{attribute}=["']#{quoted}["'][^>]*>/im
    ]
    patterns.each do |pattern|
      match = document.match(pattern)
      return CGI.unescapeHTML(match[1]).strip if match
    end
    ""
  end

  def title_tag(document)
    match = document.match(/<title[^>]*>(.*?)<\/title>/im)
    match ? CGI.unescapeHTML(match[1].gsub(/<[^>]+>/, "")).strip : ""
  end

  def absolute_http_url(value, base)
    return "" if value.empty?
    resolved = URI.join(base.to_s, value)
    %w[http https].include?(resolved.scheme) ? resolved.to_s : ""
  rescue URI::InvalidURIError
    ""
  end

  def first_non_empty(*values)
    values.find { |value| !value.to_s.empty? }.to_s
  end
end

File: backend_java/src/main/java/app/qiita/LinkPreviewService.java
package app.qiita;

import java.io.IOException;
import java.net.InetAddress;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class LinkPreviewService {
    private static final int MAX_PREVIEW_BYTES = 1_000_000;

    Map<String, Object> fetch(String rawUrl) throws Exception {
        URI current = validatePublicUrl(rawUrl);
        HttpClient previewClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(3))
            .followRedirects(HttpClient.Redirect.NEVER)
            .build();
        for (int redirects = 0; redirects <= 3; redirects++) {
            HttpRequest request = HttpRequest.newBuilder(current)
                .timeout(Duration.ofSeconds(8))
                .header("User-Agent", "Mozilla/5.0 (compatible; QiitaSearchLinkPreview/1.0)")
                .header("Accept", "text/html,application/xhtml+xml")
                .GET()
                .build();
            HttpResponse<byte[]> response;
            try {
                response = previewClient.send(request, HttpResponse.BodyHandlers.ofByteArray());
            } catch (IOException error) {
                throw new ApiException(502, "リンク先の情報を取得できませんでした。");
            }
            if (response.statusCode() >= 300 && response.statusCode() < 400) {
                String location = response.headers().firstValue("location").orElse("");
                if (location.isEmpty()) throw new ApiException(502, "リンク先の情報を取得できませんでした。");
                current = validatePublicUrl(current.resolve(location).toString());
                continue;
            }
            if (response.statusCode() < 200 || response.statusCode() >= 300) {
                throw new ApiException(502, "リンク先の情報を取得できませんでした。");
            }
            String contentType = response.headers().firstValue("content-type").orElse("").toLowerCase();
            if (!contentType.contains("text/html") && !contentType.contains("application/xhtml+xml")) {
                throw new ApiException(400, "HTMLページではないためプレビューできません。");
            }
            byte[] bytes = response.body();
            String document = new String(bytes, 0, Math.min(bytes.length, MAX_PREVIEW_BYTES), StandardCharsets.UTF_8);
            String host = current.getHost();
            String title = firstNonEmpty(meta(document, "property", "og:title"), title(document), host);
            String description = firstNonEmpty(meta(document, "property", "og:description"), meta(document, "name", "description"));
            String siteName = firstNonEmpty(meta(document, "property", "og:site_name"), host);
            return Map.of(
                "url", current.toString(),
                "title", truncate(title, 300),
                "description", truncate(description, 500),
                "image", absoluteHttpUrl(meta(document, "property", "og:image"), current),
                "site_name", truncate(siteName, 100)
            );
        }
        throw new ApiException(502, "リンク先の情報を取得できませんでした。");
    }

    private URI validatePublicUrl(String raw) {
        URI uri;
        try {
            uri = URI.create(raw);
        } catch (IllegalArgumentException error) {
            throw new ApiException(400, "http または https のURLを指定してください。");
        }
        if (!List.of("http", "https").contains(uri.getScheme()) || uri.getHost() == null) {
            throw new ApiException(400, "http または https のURLを指定してください。");
        }
        if (uri.getUserInfo() != null) {
            throw new ApiException(400, "認証情報を含むURLはプレビューできません。");
        }
        try {
            InetAddress[] addresses = InetAddress.getAllByName(uri.getHost());
            for (InetAddress address : addresses) {
                if (address.isAnyLocalAddress() || address.isLoopbackAddress() || address.isLinkLocalAddress()
                    || address.isSiteLocalAddress() || address.isMulticastAddress()) {
                    throw new ApiException(400, "ローカルネットワークのURLはプレビューできません。");
                }
            }
        } catch (IOException error) {
            throw new ApiException(400, "リンク先のホストを解決できません。");
        }
        return uri;
    }

    private static String meta(String document, String attribute, String value) {
        String quoted = Pattern.quote(value);
        List<Pattern> patterns = List.of(
            Pattern.compile("<meta[^>]+" + attribute + "=[\"']" + quoted + "[\"'][^>]+content=[\"']([^\"']*)[\"'][^>]*>", Pattern.CASE_INSENSITIVE | Pattern.DOTALL),
            Pattern.compile("<meta[^>]+content=[\"']([^\"']*)[\"'][^>]+" + attribute + "=[\"']" + quoted + "[\"'][^>]*>", Pattern.CASE_INSENSITIVE | Pattern.DOTALL)
        );
        for (Pattern pattern : patterns) {
            Matcher matcher = pattern.matcher(document);
            if (matcher.find()) return htmlDecode(matcher.group(1)).trim();
        }
        return "";
    }

    private static String title(String document) {
        Matcher matcher = Pattern.compile("<title[^>]*>(.*?)</title>", Pattern.CASE_INSENSITIVE | Pattern.DOTALL).matcher(document);
        return matcher.find() ? htmlDecode(matcher.group(1).replaceAll("<[^>]+>", "")).trim() : "";
    }

    private static String htmlDecode(String value) {
        return value.replace("&amp;", "&").replace("&quot;", "\"").replace("&#39;", "'")
            .replace("&lt;", "<").replace("&gt;", ">");
    }

    private static String absoluteHttpUrl(String value, URI base) {
        if (value.isEmpty()) return "";
        try {
            URI resolved = base.resolve(value);
            return List.of("http", "https").contains(resolved.getScheme()) ? resolved.toString() : "";
        } catch (IllegalArgumentException error) {
            return "";
        }
    }

    private static String truncate(String value, int maximum) {
        int count = value.codePointCount(0, value.length());
        return count <= maximum ? value : value.substring(0, value.offsetByCodePoints(0, maximum));
    }

    private static String firstNonEmpty(String... values) {
        for (String value : values) if (value != null && !value.isEmpty()) return value;
        return "";
    }
}
